File: libinstpatch/examples/create_sf2.py
#!/usr/bin/python

# Python example of creating a SoundFont from multiple audio samples
#
# Updated on July 22, 2006
# Now uses more convenient converter functions.
#
# Josh Green - Sep 13, 2004
# Use this example as you please (public domain)
#
# Note that the Python binding API is not stable yet and may change
#

import ipatch
import gobject, sys

if len (sys.argv[1:]) == 0:
  print "Usage: create_sf2.py sample1.wav [sample2.wav sample3.aiff ...]"
  sys.exit ()

sf2 = ipatch.SF2 ()		# Create new SoundFont object

# For each file specified on command line..
for fname in sys.argv[1:]:
  # Identify and open the file
  f = ipatch.ipatch_file_identify_open (fname, "r")

  if not f:
    print "Failed to identify file '%s'" % fname
    continue

  # Convert sample file to SF2Sample object
  sample = ipatch.ipatch_convert_object_to_type (f, ipatch.SF2Sample)
  if not sample: continue

  sf2.add_unique (sample)		# Add sample and ensure name is unique

  name = sample.get_property ("name")	# Get the sample name

  inst = ipatch.SF2Inst ()		# Create new SoundFont instrument
  inst.set_property ("name", name)	# Set the instrument name = sample name
  inst.new_zone (sample)		# Add new zone to inst => sample
  sf2.add_unique (inst)			# Add inst and ensure name is unique

  preset = ipatch.SF2Preset ()		# Create new preset
  preset.set_property ("name", name)	# Set preset name = sample name
  preset.new_zone (inst)		# Add zone to preset => inst
  sf2.add_unique (preset)		# Add preset, name/bank/preset # unique

# Create SoundFont file object, set its name and open for writing
sffile = ipatch.SF2File ()
sffile.open ("output.sf2", "w")

ipatch.ipatch_convert_objects (sf2, sffile)

sffile.close ()

File: libinstpatch/examples/create_sf2.c
/*
 * Example of creating a SoundFont from multiple audio samples
 *
 * Updated on Oct 14, 2010
 * Some API changes and additional comments in code.
 *
 * Updated on Nov 12, 2004
 * Now uses more convenient converter functions.
 *
 * Joshua "Element" Green - Sep 14, 2004
 * Use this example as you please (public domain)
 */
#include <stdio.h>
#include <libinstpatch/libinstpatch.h>

int
main (int argc, char *argv[])
{
  IpatchSF2 *sf2;
  IpatchFileHandle *fhandle;
  IpatchSF2File *sffile;
  GObject *sample;	/* IpatchSF2Sample */
  IpatchSF2Inst *inst;
  IpatchSF2Preset *preset;
  GError *err = NULL;
  char *fname, *name;
  int i;

  if (argc < 2)
  {
    fprintf (stderr, "Usage: create_sf2 sample1.wav [sample2.wav sample3.aiff ..]\n");
    return (1);
  }

  /* initialize libInstPatch */
  ipatch_init ();

  sf2 = ipatch_sf2_new ();	/* ++ ref new SoundFont object */

  /* loop over file names (command line args) */
  for (i = 1; i < argc; i++)
  {
    fname = argv[i];

    /* ++ identify file type and open handle to file object */
    fhandle = ipatch_file_identify_open (fname, &err);

    if (!fhandle)
    {
      fprintf (stderr, "Failed to identify file '%s': %s\n", fname,
               ipatch_gerror_message (err));
      g_clear_error (&err);
      continue;
    }

    /* try to convert file to a SoundFont sample */
    sample = ipatch_convert_object_to_type (G_OBJECT (fhandle->file),
                                            IPATCH_TYPE_SF2_SAMPLE, &err);
    if (!sample)
    {
      fprintf (stderr, "Failed to convert file '%s' to SoundFont sample: %s\n",
               fname, ipatch_gerror_message (err));
      g_clear_error (&err);
      ipatch_file_close (fhandle);  /* -- close file */
      continue;
    }

    /* append sample to SoundFont (ensure that its name is unique) */
    ipatch_container_add_unique (IPATCH_CONTAINER (sf2), IPATCH_ITEM (sample));

    g_object_get (sample, "name", &name, NULL); /* get the sample's name */

    /* create new SoundFont instrument (++ ref) */
    inst = ipatch_sf2_inst_new ();
    g_object_set (inst, "name", name, NULL); /* set instrument name */

    /* create new instrument zone and link sample to it */
    ipatch_sf2_inst_new_zone (inst, IPATCH_SF2_SAMPLE (sample));

    g_object_unref (sample); /* -- unref SoundFont sample */

    /* append instrument to SoundFont (ensure that its name is unique) */
    ipatch_container_add_unique (IPATCH_CONTAINER (sf2), IPATCH_ITEM (inst));

    /* create new SoundFont preset (++ ref) */
    preset = ipatch_sf2_preset_new ();
    g_object_set (preset, "name", name, NULL); /* set preset name */

    /* create new preset zone and link instrument to it */
    ipatch_sf2_preset_new_zone (preset, inst);

    g_object_unref (inst);	/* -- unref SoundFont instrument */

    /* append preset to SoundFont (ensure name/bank/preset # are unique) */
    ipatch_container_add_unique (IPATCH_CONTAINER (sf2), IPATCH_ITEM (preset));
    g_object_unref (preset);

    g_free (name);  /* free the name (returned from g_object_get) */

    ipatch_file_close (fhandle);  /* -- close file */
  }


  /* create SoundFont file object, set its name and open for writing */
  sffile = ipatch_sf2_file_new ();
  ipatch_file_set_name (IPATCH_FILE (sffile), "output.sf2");

  /* Save SoundFont to file using converter system */
  if (!ipatch_convert_objects (G_OBJECT (sf2), G_OBJECT (sffile), &err))
  {
    fprintf (stderr, "Failed to save SoundFont to file: %s\n",
             ipatch_gerror_message (err));
    g_clear_error (&err);
  }

  g_object_unref (sffile);	/* -- unref SoundFont file */
  g_object_unref (sf2);		/* -- unref SoundFont object */

  return (0);			/* we done, yeah! :) */
}
